### ISSUE TITLE
Bug in LM and DEScrypt OpenCL formats for when target mask factor is 0

### ISSUE THREAD

**Issue by magnumripper** [openwall/john#4897]
This happens when using `--rules-stack` because we need to disable internal mask then. It can (at least theoretically) also happen on very weak GPUs just from the formula that sets the target from `opencl_speed_index() / some_divisor`. The `--mask-internal-target` option added in #4879 can also be used to set the target to 0 and this is how I found this problem.

Both LM & DEScrypt formats segfaults, because they use code paths for internal mask where they shouldn't. All other formats seem to behave like they should.

My first fix:
```diff
diff --git a/src/opencl_DES_bs_plug.c b/src/opencl_DES_bs_plug.c
index 816f50735..87015409b 100644
--- a/src/opencl_DES_bs_plug.c
+++ b/src/opencl_DES_bs_plug.c
@@ -1145,7 +1145,7 @@ size_t create_keys_kernel_set_args(int mask_mode)
        cl_ulong const_cache_size;
        int i;
 
-       if (mask_mode) {
+       if (mask_mode && mask_int_cand_target) {
                fmt_opencl_DES.methods.set_key = set_key_mm;
                fmt_opencl_DES.methods.get_key = get_key_mm;
        }
diff --git a/src/opencl_lm_b_plug.c b/src/opencl_lm_b_plug.c
index a6646bbae..9af5ecae4 100644
--- a/src/opencl_lm_b_plug.c
+++ b/src/opencl_lm_b_plug.c
@@ -1175,7 +1175,7 @@ static void reset(struct db_main *db)
                release_buffer();
                create_buffer(num_loaded_hashes, offset_table, offset_table_size, hash_table_size, bitmaps, bitmap_size_bits);
 
-               if (options.flags & FLG_MASK_CHK) {
+               if (options.flags & FLG_MASK_CHK && mask_int_cand_target) {
                        mask_mode = 1;
                        fmt_opencl_lm.methods.set_key = opencl_lm_set_key_mm;
                        fmt_opencl_lm.methods.get_key = get_key_mm;
```
While I haven't *seen* any problems with this fix, it doesn't feel correct - because the `mask_mode` variable is still set and that affects other things.

So I tried changing the setting of the latter instead, but then the segfaults got back. So I dropped the above fix from #4879 and moved this problem to here.

**Comment by magnumripper** [comment]
Notes to self:
The fact that we don't have an internal mask *multiplier* doesn't necessarily mean we don't have an internal mask. Consider a mask of "?w2021". I'm not (yet) sure whether the adding of "2021" can happen on GPU side, but if it does the `mask_int_cand.num_int_cand` will still be 1.

Perhaps the above patch is a perfectly fine fix for this issue, I'm just not sure yet.

**Comment by magnumripper** [comment]
OK, the above patch is definitely not correct for DEScrypt. I ran a test for a mask of "?w1" and it didn't find a single password when using `--mask-internal-target=0`.

```
$ grep -v '^#!comment' ../run/password.lst | grep -P '^.{2,7}1$' | ../run/pass_gen.pl descrypt > destest.in
$ sed 's/.$//' ../run/password.lst > passwor.lst
$ make -sj8 && rm -f ../run/john.pot && ../run/john -form:descrypt-opencl -dev=cpu destest.in -w:passwor.lst -mask:?w1 --mask-internal-target=0
```
It's supposed to find 196g.

For LM,
```
$ grep -v '^#!comment' ../run/password.lst | grep -P '^.{2,6}1$' | ../run/pass_gen.pl lm > lmtest.in
$ make -sj8 && rm -f ../run/john.pot && ../run/john -form:lm-opencl -dev=cpu lmtest.in -w:passwor.lst -mask:?w1 --mask-internal-target=0
```
is supposed to find 164g. For this test, LM-opencl segfaults patched or not, and using `--mask-internal-target=0` or not...

**Comment by magnumripper** [comment]
Here's a promising fix (while not necessarily complete)
```diff
diff --git a/src/mask_ext.c b/src/mask_ext.c
index 87d7a7083..6bd017460 100644
--- a/src/mask_ext.c
+++ b/src/mask_ext.c
@@ -128,9 +128,6 @@ void mask_ext_calc_combination(mask_cpu_context *ptr, int max_static_range)
        mask_int_cand.int_cpu_mask_ctx = NULL;
        mask_int_cand.int_cand = NULL;
 
-       if (!mask_int_cand_target)
-               return;
-
        if (MASK_FMT_INT_PLHDR > 4) {
                fprintf(stderr, "MASK_FMT_INT_PLHDR value must not exceed 4.\n");
                error();
```

**Comment by magnumripper** [comment]
~That last change actually seems to be the fix!~

Nope, the reason it works is it ends up with eg. x10 on GPU side, instead of all CPU-side.

**Comment by magnumripper** [comment]
While throwing stones at this to see what what sticks, I notice the problems and partial fixes are slightly differerent not only between LM and DEScrypt but also between macOS DEScrypt and Linux NVIDIA DEScrypt.

It seems `bs_b` is the "basic kernel", `bs_h` is "hardcoded salt" and `bs_f` is "full unroll" (and also has hardcoded salt). For LM, there is no `bs_h`, there's just "basic" and "full unrolled".

opencl_DES_fmt_plug.c
```c
#define USE_FULL_UNROLL         (amd_gcn(device_info[gpu_id]) || nvidia_sm_5x(device_info[gpu_id]))                     
#define USE_BASIC_KERNEL        (cpu(device_info[gpu_id]) || platform_apple(platform_id))                               

(...)

	if ((USE_BASIC_KERNEL&& !OVERRIDE_AUTO_CONFIG) ||
		(OVERRIDE_AUTO_CONFIG && !HARDCODE_SALT && !FULL_UNROLL))
		opencl_DES_bs_b_register_functions(pFmt);
	else if ((USE_FULL_UNROLL && !OVERRIDE_AUTO_CONFIG) ||
		(OVERRIDE_AUTO_CONFIG && HARDCODE_SALT && FULL_UNROLL))
		opencl_DES_bs_f_register_functions(pFmt);
	else
		opencl_DES_bs_h_register_functions(pFmt);
```
...errr so `bs_h` is used if neither AMD GPU, NVIDIA Maxwell (not newer ones?), CPU or macOS. That's an unrelated bug right there. If I do this:
```diff
-#define USE_FULL_UNROLL                (amd_gcn(device_info[gpu_id]) || nvidia_sm_5x(device_info[gpu_id]))
+#define USE_FULL_UNROLL                (amd_gcn(device_info[gpu_id]) || nvidia_sm_5x(device_info[gpu_id]) || nvidia_sm_6x(device_info[gpu_id]) || nvidia_sm_7x(device_info[gpu_id]))
```
...the DEScrypt speed on a 2080ti goes from
```
Many salts:     1190M c/s real, 1184M c/s virtual, Dev#2 util: 100%
Only one salt:  1134M c/s real, 1134M c/s virtual, Dev#2 util: 99%
```
to
```
Many salts:     1895M c/s real, 1876M c/s virtual, Dev#2 util: 100%
Only one salt:  1745M c/s real, 1745M c/s virtual, Dev#2 util: 98%
```
It stops building per-salt kernels, to boot!

**Comment by solardiz** [comment]
In general, I think for then-latest GPUs we should use NVIDIA compute capability (and similar for other vendors) checks of the kind ">= revision", not "== revision". In other words, I think the best guess we can make is that future GPUs will be similar to latest current, not to some arbitrary older. Similarly, we should now fix this check not by listing more `nvidia_sm_*`, but by introducing and using a new macro like `nvidia_sm_5x_plus`.

**Comment by solardiz** [comment]
> It stops building per-salt kernels, to boot!

Hmm, we don't use per-salt descrypt kernels on AMD and recent NVIDIA? That's weird! The per-salt kernels optimization is expected to be universal. It would in fact be nice to be able to manually disable it for faster startup, though.

**Comment by magnumripper** [comment]
It can be disabled by editing `run/opencl/opencl_DES_hst_dev_shared.h` and set `OVERRIDE_AUTO_CONFIG` and other params accordingly. Even though that file is in the kernels directory, a full rebuild of the host code is needed as far as I can see.

Now that I finally sorted this out, it would be trivial to implement some option for forcing `bs_b` at run-time.

**Comment by magnumripper** [comment]
I think I'll at least introduce an environment variable eg. `DES_KERNEL` that can be set to `bs_b`, `bs_h` or `bs_f`. Without that it's hard to fix this issue's bugs and test the fixes.

**Comment by solardiz** [comment]
I suggest we prefix our env var names with `JOHN_`, perhaps except only for `LWS` and `GWS`. To avoid inadvertent usage.

**Comment by magnumripper** [comment]
> It stops building per-salt kernels, to boot!

That really doesn't make sense though. Both bf_h and bf_f use per-salt kernels, only bf_b, the basic kernel used for CPU's and stuff, is a single kernel for all salts.
```
$ git grep "per-salt" opencl*c
opencl_DES_bs_f_plug.c:                 fprintf(stderr, "Note: Building per-salt kernels. This takes e.g. 2 hours for 4096 salts.\n");
opencl_DES_bs_h_plug.c:                 fprintf(stderr, "Note: Building per-salt kernels. This takes e.g. 2 hours for 4096 salts.\n");
```
I did see it stop building salts, and I did double-check it. My mind is a mess now with all these little issues that are entangled. I need to sort out one thing at a time.

**Comment by solardiz** [comment]
> I did see it stop building salts, and I did double-check it.

Maybe that's because the kernels for those salts were already built and cached by previous runs?

**Comment by magnumripper** [comment]
The "2 hours" message pops up anyway, cached kernels just make it a lot faster.

But never mind, what I do see now is that on the 2080ti, bs_f is faster than bs_h while bs_b fails with "CL_OUT_OF_RESOURCES (-5) error in opencl_DES_bs_plug.c:544 - Enque kernel DES_bs_cmp failed". I'll postpone looking into why, and just make a PR for the selection mechanisms.

**Comment by magnumripper** [comment]
New start on this at cf5536e7b535 in the main tree.

This on Linux / 2080ti

Things to check:
- [ ] descrypt running bs_b with just --test: `CL_OUT_OF_RESOURCES (-5) error in opencl_DES_bs_b_plug.c:687 - Enque kernel DES_bs_25 failed`
- [ ] descrypt running bs_b with just --test --no-mask: `CL_INVALID_KERNEL_ARGS (-52) error in opencl_DES_bs_b_plug.c:687 - Enque kernel DES_bs_25 failed`
- [x] descrypt running bs_f with just --test
- [x] descrypt running bs_h with just --test
- [x] lm running bs_b with just --test
- [x] lm running bs_f with just --test
- [x] descrypt running bs_b with normal (accelerating) mask
- [x] descrypt running bs_f with normal (accelerating) mask
- [x] descrypt running bs_h with normal (accelerating) mask
- [x] lm running bs_b with normal (accelerating) mask
- [x] lm running bs_f with normal (accelerating) mask
- [ ] descrypt running bs_b with a x1 mask such as "?w1": `CL_OUT_OF_RESOURCES (-5) error in opencl_DES_bs_plug.c:544 - Enque kernel DES_bs_cmp failed`
- [x] descrypt running bs_f with a x1 mask such as "?w1"
- [x] descrypt running bs_h with a x1 mask such as "?w1"
- [ ] lm running bs_b with a x1 mask such as "?w1": `CL_INVALID_COMMAND_QUEUE (-36) error in opencl_lm_b_plug.c:1305 - Kernel failed`
- [ ] lm running bs_f with a x1 mask such as "?w1": `CL_INVALID_COMMAND_QUEUE (-36) error in opencl_lm_b_plug.c:1305 - Kernel failed`

After 63f451b8c1:
- [ ] descrypt running bs_b with just --test: `CL_OUT_OF_RESOURCES (-5) error in opencl_DES_bs_plug.c:544 - Enqueue kernel DES_bs_cmp failed`
- [ ] descrypt running bs_b with just --test --no-mask: `CL_INVALID_KERNEL_ARGS (-52) error in opencl_DES_bs_b_plug.c:687 - Enque kernel DES_bs_25 failed`
- [x] descrypt running bs_f with just --test
- [x] descrypt running bs_h with just --test
- [x] lm running bs_b with just --test
- [x] lm running bs_f with just --test
- [x] descrypt running bs_b with normal (accelerating) mask
- [x] descrypt running bs_f with normal (accelerating) mask
- [x] descrypt running bs_h with normal (accelerating) mask
- [x] lm running bs_b with normal (accelerating) mask
- [x] lm running bs_f with normal (accelerating) mask
- [ ] descrypt running bs_b with a x1 mask such as "?w1": `CL_OUT_OF_RESOURCES (-5) error in opencl_DES_bs_plug.c:544 - Enqueue kernel DES_bs_cmp failed`
- [ ] descrypt running bs_f with a x1 mask such as "?w1": `CL_OUT_OF_RESOURCES (-5) error in opencl_DES_bs_plug.c:544 - Enqueue kernel DES_bs_cmp failed`
- [ ] descrypt running bs_h with a x1 mask such as "?w1": `CL_OUT_OF_RESOURCES (-5) error in opencl_DES_bs_plug.c:544 - Enqueue kernel DES_bs_cmp failed`
- [ ] lm running bs_b with a x1 mask such as "?w1": `CL_INVALID_COMMAND_QUEUE (-36) error in opencl_lm_b_plug.c:1305 - Kernel failed`
- [ ] lm running bs_f with a x1 mask such as "?w1": `CL_INVALID_COMMAND_QUEUE (-36) error in opencl_lm_b_plug.c:1305 - Kernel failed`

**Comment by magnumripper** [comment]
So what I thought was correct was in fact the opposite.

After 2529a9e97 (or 7fcbda4c9 as it were after merging)

Things to check:
- [ ] descrypt running bs_b with just --test: `CL_OUT_OF_RESOURCES (-5) error in opencl_DES_bs_b_plug.c:687 - Enqueue kernel DES_bs_25 failed`
- [ ] descrypt running bs_b with just --test --no-mask: `CL_INVALID_KERNEL_ARGS (-52) error in opencl_DES_bs_b_plug.c:687 - Enqueue kernel DES_bs_25 failed`
- [x] descrypt running bs_f with just --test
- [x] descrypt running bs_h with just --test
- [x] lm running bs_b with just --test
- [x] lm running bs_f with just --test
- [x] descrypt running bs_b with normal (accelerating) mask
- [x] descrypt running bs_f with normal (accelerating) mask
- [x] descrypt running bs_h with normal (accelerating) mask
- [x] lm running bs_b with normal (accelerating) mask
- [x] lm running bs_f with normal (accelerating) mask
- [ ] descrypt running bs_b with a x1 mask such as "?w1": `CL_INVALID_KERNEL_ARGS (-52) error in opencl_DES_bs_b_plug.c:687 - Enqueue kernel DES_bs_25 failed`
- [x] descrypt running bs_f with a x1 mask such as "?w1"
- [x] descrypt running bs_h with a x1 mask such as "?w1"
- [x] lm running bs_b with a x1 mask such as "?w1"
- [x] lm running bs_f with a x1 mask such as "?w1"

So now only DEScrypt-opencl using basic kernel is problematic. Phew.

**Comment by magnumripper** [comment]
It's weird that a simple `--test` fails (even crashes) for DEScrypt-opencl using `bs_b` while a "real test" using a normal accelerating mask works fine (and cracks what it should). They should be the same thing. I tried to look for intermittent failures but haven't seen evidence of it yet.

**Comment by magnumripper** [comment]
The issue as posted in OP seems to be fixed with 2529a9e (or 7fcbda4 as it were after merging). The remaining problem with DEScrypt `bs_b` is a different issue, #4901
